Add instructions for manually configuring an Azure Batch pool (#325)
seqeralabs/docs: platform_versioned_docs/version-24.2/compute-envs/azure-batch.mdx

    - Add the **Batch account** and **Blob Storage account** names and access keys to the relevant fields.
1. Delete the copied keys from their temporary location after they have been added to a credential in Platform.

#### Entra service principal and managed identity

> If using Entra for authentication, you must create a service principal and managed identity. Seqera Platform uses the Service Principal to authenticate to Azure Batch and Azure Storage. It submits a Nextflow task as the head process to run Nextflow, which authenticates to Azure Batch and storage using the Managed Identity attached to the node pool.

Reviewer: ❓ Should there be links out to MSFT-managed docs explaining SPs? Oh I see, it's described in a section above. Point still holds true (just for the earlier description).  Nevermind, I see you have it below.

⛏️ I'd probably frontload the explanations, but this works too.

> If using Entra for authentication, you must create a service principal and managed identity. Seqera Platform uses the Service Principal to authenticate to Azure Batch and Azure Storage. It submits a Nextflow task as the head process to run Nextflow, which authenticates to Azure Batch and storage using the Managed Identity attached to the node pool.

Reviewer: ❓ SP type qualification necessary?  (i.e. user-managed). Nevermind, I see this is handled below.

> If using Entra for authentication, you must create a service principal and managed identity. Seqera Platform uses the Service Principal to authenticate to Azure Batch and Azure Storage. It submits a Nextflow task as the head process to run Nextflow, which authenticates to Azure Batch and storage using the Managed Identity attached to the node pool.

Reviewer: ❓ Do our docs adequately cover how to attach an SP to an Azure Batch pool (I haven't checked)? If yes, does this cover both Manual and Forge flows (assuming both are applicable)?

> Therefore, you must create both an Entra service principal and a managed identity. You add the service principal to your Seqera Platform credentials and attach the managed identity to your Azure Batch node pool which will run Nextflow.

Reviewer: If I understand this correctly, two different MSFT identities need to be used:

1. Service Principle type -- which can accommodate the fact that Tower lives outside the Azure network and likely is calling across the public internet.

2. Managed Identity -- can be used since the Pool lives inside the account and thus has a greater assurance level.

Correct?

Yes, explained below:
> When you use a manually configured compute environment with a managed identity attached to the Azure Batch Pool, Nextflow can use this managed identity for authentication. However, Platform still needs to use access keys or an Entra service principal to submit the initial task to Azure Batch to run Nextflow, which will then proceed with the managed identity for subsequent authentication.

1. **Family**: The first letter of the VM size name indicates the machine family. For example, `Standard_E16d_v5` is a member of the E family.
    - *A*: Economical machines, low power machines.
    - *B*: Burstable machines which use credits for cost allocation.
    - *D*: General purpose machines suitable for most applications.
    - *DC*: D machines with additional confidential compute capabilities.
    - *E*: The same as D but with more memory. These are generally the best machines for bioinformatics workloads.
    - *EC*: The same as E but with additional confidential compute capabilities.
    - *F*: Compute optimized machines which come with a faster CPU compared to D-series machines.
    - *M*: Memory optimized machines which come with extremely large and fast memory layers, typically more than is needed for bioinformatics workloads.
    - *L*: Storage optimized machines which come with large locally attached NVMe storage drives. Note that these need to be configured before you can use them with Azure Batch.

>     - *N*: Accelerated computing machines which come with FPGAs, GPUs or custom ASICs.

Reviewer 2: ```suggestion
    - *N*: Accelerated computing machines which come with FPGAs, GPUs, or custom ASICs.
```

    - *H*: High performance machines which come with the fastest processors and memory

In general, we recommend using the E family of machines for bioinformatics workloads since these are cost effective, widely available and sufficiently fast.

> 1. **vCPUs**: The number of vCPUs the machine has. This is the main factor in determining the speed of the machine.

Reviewer: 🤔 Personally I like this level of detail, but there is a risk of it getting out of date. Docs probably will need to set themselves a recurring task to go check accuracy every so often.

See [Launch pipelines](../launch/launchpad.mdx) to start executing workflows in your Azure Batch compute environment.
:::

> ### Manual

Reviewer: ❓ I don't see Forge here. Is that an omission by accident, or can the SP / Managed Identity config only occur via Manual creation for now?

> It is possible to set up Seqera Platform to use a pre-existing Azure Batch pool. This allows the use of more advanced Azure Batch features, such as custom VM images and private networking.

Reviewer: Should we link out to MSFT pages that explain the benefits of why you might want these features?

1. **VM size**: This is the size of the VM. See [the section on Azure VM sizes][az-vm-sizes] for more information.
1. **Scale**: Azure Node pools can be fixed in size or autoscale based on a formula. We recommend autoscaling to enable scaling your resources down to zero when not in use. Click **Auto scale**. Change the **AutoScale evaluation interval** to 5 minutes, this is the minimum period between evaluations of the autoscale formula. For formula, you can use any valid formula, please see the documentation [here][az-batch-autoscale] for more information. This is the default autoscaling formula, with a maximum of 8 VMs:

```
// Compute the target nodes based on pending tasks.
// $PendingTasks == The sum of $ActiveTasks and $RunningTasks
$samples = $PendingTasks.GetSamplePercent(interval);
$tasks = $samples < 70 ? max(0, $PendingTasks.GetSample(1)) : max( $PendingTasks.GetSample(1), avg($PendingTasks.GetSample(interval)));
$targetVMs = $tasks > 0 ? $tasks : max(0, $TargetDedicatedNodes/2);
targetPoolSize = max(0, min($targetVMs, 8));

// For first interval deploy 1 node, for other intervals scale up/down as per tasks.
$TargetDedicatedNodes = targetPoolSize;

> $NodeDeallocationOption = taskcompletion;

Reviewer 3: With this I was able to create the pool.

$TargetLowPriorityNodes = lifespan < interval ? 1 : targetPoolSize;
$NodeDeallocationOption = taskcompletion;
```

1. **Start task**: This is the task that will run on each VM when it joins the pool. This can be used to install additional software on the VM. When using Batch Forge, this is used to install `azcopy` for staging files onto and off the node. Select **Enabled** and add the following command line to install `azcopy`:

```shell
bash -c "chmod +x azcopy && mkdir $AZ_BATCH_NODE_SHARED_DIR/bin/ && cp azcopy $AZ_BATCH_NODE_SHARED_DIR/bin/"
```

Click **Resource files** then select **Http url**. For the URL, add `https://nf-xpack.seqera.io/azcopy/linux_amd64_10.8.0/azcopy` and for File path type `azcopy`. Every other setting can be left default.

    :::note
    When not using Fusion, every node **must** have `azcopy` installed.
    :::

1. **Task Slots**: Set task slots to the number of vCPUs the machine has, e.g. select `4` for a `Standard_D4_v3` VM size.
1. **Task scheduling policy**: This can be set to `Pack` or `Spread`. `Pack` will attempt to schedule tasks from the same job on the same VM, while `Spread` will attempt to distribute tasks evenly across VMs.

> 1. **Virtual Network**: If using a virtual network, you can select it here. Be sure to select the correct virtual network and subnet. The virtual machines require:

Reviewer 2: ```suggestion
1. **Virtual Network**: If you are using a virtual network, you can select it here. Be sure to select the correct virtual network and subnet. The VMs require:
```

    - Access to container registries (e.g. quay.io, docker.io) to pull containers
    - Access to Azure Storage to copy data using `azcopy`
    - Access to any remote files required by the pipeline e.g. AWS S3.
    - Communication with the head node (running Nextflow) and Seqera Platform to relay logs and information
Note that overly restrictive networking may prevent pipelines from running successfully.
1. **Mount configuration**: Nextflow *only* supports Azure File Shares. Select `Azure Files Share`, then add:
    - **Source**: URL in format `https://${accountName}.file.core.windows.net/${fileShareName}`
    - **Relative mount path**: Path where the file share will be mounted on the VM
    - **Storage account name** and **Storage account key** (managed identity is not supported)
    
Leave the node pool to start and create a single Azure VM. Monitor the VM to ensure it starts correctly. If any errors occur, check and correct them - you may need to create a new Azure node pool if issues persist.

The following settings can be modified after creating a pool:

- Autoscale formula

> - Start Task

Reviewer 2: ```suggestion
- Start task
```

:::

### Manual

You can configure Seqera Platform to use a pre-existing Azure Batch pool. This allows the use of more advanced Azure Batch features, such as custom VM images and private networking. See [Azure Batch security best practices][az-batch-best-practices] for more information.

:::caution
Your Seqera compute environment uses resources that you may be charged for in your Azure account. See [Cloud costs](../monitoring/cloud-costs.mdx) for guidelines to manage cloud resources effectively and prevent unexpected costs.
:::

#### Create a Nextflow-compatible Azure Batch pool

If not described below, use the default settings:

1. **Account**: You must have an existing Azure Batch account. Ideally, you should already have tested whether you can run an Azure Batch task within this account. Any type of account is compatible.
1. **Quota**: You must check you have sufficient quota for the number of pools, jobs, and vCPUs per series. See [Azure Batch service quotas and limits][az-batch-quotas] for more information.
1. On the Azure Batch page of the Azure Portal, select **Pools** and then **+ Add**.

> 1. **Name**: Enter a Pool ID and Display Name. This ID will be used by Seqera and Nextflow.

Reviewer 2: ```suggestion
1. **Name**: Enter a **Pool ID** and **Display Name**. This ID will be used by Seqera and Nextflow.
```

1. **Identity**: Select **User assigned** to use a managed identity for the pool. Select **Add** for the user-assigned managed identity and select the managed identity with the correct permissions to the Azure Storage and Batch accounts.
1. **Operating System**: You can use any Linux-based image here, but it is recommended to use it with a Microsoft Azure Batch-provided image. Note that there are two generations of Azure Virtual Machine images, and certain VM series are only available in one generation. See [Azure Virtual Machine series][az-vm-gen] for more information. For default settings, select the following:

>     - **Publisher**: `microsoft-azure-batch`

Reviewer 2: Just to confirm: Are these field entries or CLI? If they're in a UI field they don't need to be code format.

Reviewer 2: That makes sense, thank you!

    Configuration settings in this field override the same values in the pipeline repository `nextflow.config` file. See [Nextflow config file](../launch/advanced.mdx#nextflow-config-file) for more information on configuration priority. 
    :::
    :::info

>     To use managed identities, Platform requires Nextflow version 24.06.0-edge or later.

Reviewer: ⛏️ Does "Platform" really require it? If Managed Identity is being leveraged by the head job in the Batch Pool, isn't this more a matter of "The job your launch via Tower into the Managed-Identity-Pool must use Nextflow 24.06.0-edge or greater"?

    Configuration settings in this field override the same values in the pipeline repository `nextflow.config` file. See [Nextflow config file](../launch/advanced.mdx#nextflow-config-file) for more information on configuration priority. 
    :::
    :::info

>     Nextflow must be version 24.10.0 or later to use managed identities.

Reviewer 4: @adamrtalbot This contradicts the earlier admonitions that states 24.06.0-edge is required. This admonition also seems to be in the wrong place, as nothing relates to managed identities here
